source/bigint.c

```diff
@@ -0,0 +1,353 @@
+
+        hex_digits.len -= digit_cursor.len;
+    }
+
+    bigint->sign = 1;
+
+    return AWS_OP_SUCCESS;
+
+on_error:
+
+    aws_bigint_clean_up(bigint);
+
+    return AWS_OP_ERR;
+}
+
+int aws_bigint_init_from_int64(struct aws_bigint *bigint, struct aws_allocator *allocator, int64_t value) {
+    if (value >= 0) {
+        return aws_bigint_init_from_uint64(bigint, allocator, (uint64_t)value);
+    }
+
+    if (value == INT64_MIN) {
+        if (aws_bigint_init_from_uint64(bigint, allocator, INT64_MIN_AS_HEX)) {
+            return AWS_OP_ERR;
+        }
+    } else {
+        if (aws_bigint_init_from_uint64(bigint, allocator, (uint64_t)(-value))) {
+            return AWS_OP_ERR;
+        }
+    }
+
+    bigint->sign = -1;
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_bigint_init_from_uint64(struct aws_bigint *bigint, struct aws_allocator *allocator, uint64_t value) {
+
+    uint32_t lower_digit = (uint32_t)(value & LOWER_32_BIT_MASK);
+    uint32_t upper_digit = (uint32_t)((value >> 32) & LOWER_32_BIT_MASK);
+    uint64_t digit_count = upper_digit > 0 ? 2 : 1;
+    if (aws_array_list_init_dynamic(&bigint->digits, allocator, digit_count, sizeof(uint32_t))) {
+        return AWS_OP_ERR;
+    }
+
+    aws_array_list_push_back(&bigint->digits, &lower_digit);
+
+    if (upper_digit > 0) {
+        aws_array_list_push_back(&bigint->digits, &upper_digit);
+    }
+
+    bigint->sign = 1;
+
+    return AWS_OP_SUCCESS;
+}
+
+static const uint8_t *HEX_CHARS = (const uint8_t *)"0123456789abcdef";
+
+static void s_append_uint32_as_hex(struct aws_byte_buf *buffer, uint32_t value, bool prepend_zeros) {
+
+    bool have_seen_non_zero_nibble = false;
+    size_t write_index = buffer->len;
+    for (size_t i = 0; i < NIBBLES_PER_DIGIT; ++i) {
+        uint8_t high_nibble = (uint8_t)(value >> 28);
+        AWS_FATAL_ASSERT(high_nibble < 16);
+
+        if (high_nibble > 0) {
+            have_seen_non_zero_nibble = true;
+        }
+
+        if (have_seen_non_zero_nibble || prepend_zeros || i + 1 == NIBBLES_PER_DIGIT) {
+            AWS_FATAL_ASSERT(write_index < buffer->capacity);
+            buffer->buffer[write_index++] = HEX_CHARS[high_nibble];
+        }
+
+        value <<= 4;
+    }
+
+    buffer->len = write_index;
+}
+
+int aws_bigint_bytebuf_append_as_hex(struct aws_bigint *bigint, struct aws_byte_buf *buffer) {
+    size_t digit_count = aws_array_list_length(&bigint->digits);
+    uint64_t max_hex_digits = aws_array_list_length(&bigint->digits) * NIBBLES_PER_DIGIT;
+    uint64_t total_characters = max_hex_digits + ((bigint->sign < 0) ? 1 : 0);
+    if (aws_byte_buf_reserve_relative(buffer, total_characters)) {
+        return AWS_OP_ERR;
+    }
+
+    /*
+     * We don't support negative hex numbers from an initialization standpoint, but we still
+     * need to indicate the number's sign on output
+     */
+    if (bigint->sign < 0) {
+        buffer->buffer[buffer->len++] = '-';
+    }
```

Comment:
Or we could let user write '-' themselves, just like they need to call negate() themselves when initializing.
Also gives flexibility for them to format something like "-0x01".

Reply:
Deferring output functionality decisions by downgrading this to a (temporary) debug function.